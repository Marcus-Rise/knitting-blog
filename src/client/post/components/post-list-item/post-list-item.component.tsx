import type { FC } from "react";
import React from "react";
import Image from "next/image";
import NextLink from "next/link";
import { DateToString } from "../../../../common/utils/date-to-string";
import { ImageView } from "../../../components";
import {
  Center,
  Description,
  ImageStyled,
  Label,
  Link,
  Meta,
  Title,
} from "./post-list-item.styles";
import type { IPostWithoutContent } from "../../../../common/post";

interface IProps extends IPostWithoutContent {
  className?: string;
  priority?: boolean;
}

const PostListItem: FC<IProps> = ({
  className,
  date,
  description,
  imageLabel,
  imageSrc,
  slug,
  title,
  priority,
}) => (
  <div className={className}>
    <Title>{title}</Title>
    <Meta>{DateToString(date)}</Meta>
    <ImageView album={[{ src: imageSrc, alt: imageLabel }]}>
      <ImageStyled>
        <Image
          src={imageSrc}
          alt={imageLabel}
          layout={"fill"}
          placeholder={"blur"}
          priority={priority}
          blurDataURL={
            "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNcXw8AAeMBMA+N6mYAAAAASUVORK5CYII="
          }
        />
      </ImageStyled>
    </ImageView>
    <Label>{imageLabel}</Label>
    <Description>{description}</Description>
    <Center>
      <Link>
        <NextLink
          href={{
            pathname: "/[slug]",
            query: { slug },
          }}
        >
          Читать далее
        </NextLink>
      </Link>
    </Center>
  </div>
);

export { PostListItem };
