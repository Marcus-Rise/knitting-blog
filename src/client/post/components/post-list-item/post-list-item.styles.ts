import styled from "styled-components";
import { BadScript, Bold, colors } from "../../../styles";

const Title = styled.h2`
  ${Bold};
  font-size: 2rem;
  margin-bottom: 0;
`;

const Meta = styled.p`
  color: ${colors.secondary};
  font-size: 1rem;
  text-transform: capitalize;
`;

const ImageStyled = styled.div`
  height: 320px;

  width: auto;
  overflow: hidden;

  & > div {
    &,
    & > img {
      position: relative !important;
    }
  }

  &:hover {
    cursor: pointer;
  }

  img {
    object-fit: cover;
    height: 320px !important;
    max-height: none !important;
    min-height: 0 !important;
    margin: 0 !important;
  }
`;

const Label = styled.p`
  ${BadScript};

  color: ${colors.primary};
  text-align: center;
  font-size: 1.5rem;
  margin-bottom: 0;
`;

const Description = styled.p`
  line-height: 1.75rem;
`;

const Link = styled.div`
  a {
    text-decoration: none;
    color: ${colors.primary};
    border: 0.15rem dashed ${colors.accent};
    padding: 0.5rem 3rem;
    box-sizing: border-box;
    display: block;

    &:hover {
      border: none;
      color: white;
      background-color: ${colors.accent};
      padding: 0.65rem 3rem;
    }

    &:focus {
      box-shadow: inset 0 0 4px 2px rgba(0, 0, 0, 0.25);
    }
  }
`;

const Center = styled.div`
  display: flex;
  justify-content: center;
`;

export { Title, Meta, ImageStyled, Label, Description, Link, Center };
