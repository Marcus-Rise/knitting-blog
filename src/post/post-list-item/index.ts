export { PostListItem } from "./post-list-item.component";
