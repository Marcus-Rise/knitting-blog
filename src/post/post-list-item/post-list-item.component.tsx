import React from "react";
import styles from "./post-list-item.module.scss";
import type { IPost } from "../post.interface";
import Image from "next/image";

interface IProps extends IPost {
  className?: string;
}

const PostListItem: React.FC<IProps> = (props) => (
  <div className={[props.className, styles.root].join(" ")}>
    <h2 className={styles.title}>{props.title}</h2>
    <p className={styles.meta}>{props.date}</p>
    <div className={styles.image}>
      <Image src={props.imageSrc} alt={props.imageLabel} height={320} width={"auto"} layout={"responsive"} />
    </div>
    <p className={styles.imageLabel}>{props.imageLabel}</p>
    <p className={styles.description}>{props.description}</p>
    <div className="row justify-content-center">
      <div className="col-auto">
        <a className={styles.link} href={"/" + props.slug}>
          Читать далее
        </a>
      </div>
    </div>
  </div>
);

export { PostListItem };
