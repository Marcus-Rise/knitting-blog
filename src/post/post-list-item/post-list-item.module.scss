@use "../../styles/typography";
@use "../../styles/grid";

.root {

}

.title {
  @include typography.bold;
  font-size: 2rem;
  margin-bottom: 0;
}

.meta {
  color: var(--secondary);
  font-size: 1rem;
  text-transform: capitalize;
}

.image {
  $height: 320px;
  height: $height;

  width: auto;
  overflow: hidden;

  img {
    object-fit: cover;
    height: $height !important;
    max-height: none !important;
    min-height: 0 !important;
    margin: 0 !important;
  }
}

.imageLabel {
  @include typography.BadScript;

  color: var(--primary);
  text-align: center;
  font-size: 1.5rem;
}

.description {
  line-height: 1.75rem;
}

.link {
  a {
    text-decoration: none;
    color: var(--primary);
    border: .15rem dashed var(--accent);
    padding: .5rem 3rem;
    box-sizing: border-box;
    display: block;

    &:hover {
      border: none;
      color: white;
      background-color: var(--accent);
      padding: .65rem 3rem;
    }

    &:focus {
      box-shadow: inset 0 0 4px 2px rgba(0, 0, 0, 0.25);
    }
  }
}

