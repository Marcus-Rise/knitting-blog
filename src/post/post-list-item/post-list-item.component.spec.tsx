import { render } from "@testing-library/react";
import React from "react";
import { PostListItem } from "./post-list-item.component";

describe("PostListItem", () => {
  test("render", () => {
    const { asFragment } = render(
      <PostListItem
        title={"title"}
        slug={"slug"}
        description={"description"}
        date={"2020-11-11T17:47:51+0000"}
        imageLabel={"image"}
        imageSrc={"http://test.com"}
        content={[]}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
