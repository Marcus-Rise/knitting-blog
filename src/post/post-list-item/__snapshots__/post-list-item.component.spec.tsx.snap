// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListItem render 1`] = `
<DocumentFragment>
  <div
    class=" root"
  >
    <h2
      class="title"
    >
      title
    </h2>
    <p
      class="meta"
    >
      ноябрь 11, 2020
    </p>
    <div
      data-testid="image-view"
    >
      <div
        class="image"
      >
        <img
          alt="image"
          height="auto"
          src="http://test.com"
          width="auto"
        />
      </div>
    </div>
    <p
      class="imageLabel"
    >
      image
    </p>
    <p
      class="description"
    >
      description
    </p>
    <div
      class="row justify-content-center"
    >
      <div
        class="col-auto"
      >
        <div
          class="link"
        >
          Читать далее
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
